### PR TITLE
keyword rest service

### PR DESCRIPTION
**Pull request by hbaptiste**
keyword rest service

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hi @hbaptiste, still not good: hold on ;)

File: NestedNode/KeyWord.php [hunk 1]
```diff
@@ -265,7 +265,23 @@ public function toStdObject()
         $object->level = $this->getLevel();
         $object->keyword = $this->getKeyword();
         $object->children = array();
-
         return $object;
     }
+
+    /**
+     *
```
Comment (mickaelandrieu) [inline NestedNode/KeyWord.php]: Add a comment or remove it please.

File: Rest/Controller/KeywordController.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+use BackBee\NestedNode\MediaFolder;
+use BackBee\Rest\Controller\Annotations as Rest;
+use BackBee\Rest\Patcher\EntityPatcher;
+use BackBee\Rest\Patcher\Exception\InvalidOperationSyntaxException;
+use BackBee\Rest\Patcher\Exception\UnauthorizedPatchOperationException;
+use BackBee\Rest\Patcher\RightManager;
+use BackBee\Rest\Patcher\OperationSyntaxValidator;
+use BackBee\Utils\StringUtils;
+
+/*
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: to be removed

File: NestedNode/Repository/KeyWordRepository.php [hunk 3]
```diff
@@ -51,6 +51,7 @@ public function getLikeKeyWords($cond)
             return;
         }
     }
+
```
Comment (mickaelandrieu) [inline NestedNode/Repository/KeyWordRepository.php]: unused line, remove it please.

File: Rest/Controller/KeywordController.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+ * To change this template, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+/**
+ * Description of KeywordController
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: probably generated, I'm in favor or a real "explanation" of what is done here.

For instance "Handle the Keyword management system, manage insertions, deletions and update of Keyword through a REST Api"

File: Rest/Controller/KeywordController.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * @author h.baptiste
+ */
+class KeywordController extends AbstractRestController
+{
+
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: remove all unused lines please

File: Rest/Controller/KeywordController.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+     * @param Request $request
+     *
+     * @Rest\Pagination(default_count=25, max_count=100)
+     */
+    public function getCollectionAction(Request $request, $start)
+    {
+        $queryParams = $request->query->all();
+        
+        $term = array_key_exists("term", $queryParams) ? $queryParams["term"] : null;
+        if(null === $term) {
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: a space after `if` statment please. You'd better use php cs fixer in this file to be sure the code is PSR compliant.

File: Rest/Controller/KeywordController.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+   /**
+     * @param Request $request
+     *
+     * @Rest\Pagination(default_count=25, max_count=100)
+     */
+    public function getCollectionAction(Request $request, $start)
+    {
+        $queryParams = $request->query->all();
+        
+        $term = array_key_exists("term", $queryParams) ? $queryParams["term"] : null;
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: nope.

```php
$term = $request->query->get('term', null);
```

File: Rest/Controller/KeywordController.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        if(null === $term) {
+            throw new BadRequestHttpException('A term should be provided.');
+        }
+        $paginator = $this->getKeywordRepository()->getLikeKeyWords($term);
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: if you need an array, create a new method in KeyWordRepository that return the array you need.

To be honest, the Symfony/BackBee way is "don't do any business in Controllers, only get and return things"

File: Rest/Controller/KeywordController.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+        $lastResult = $start + $count - 1;
+        $lastResult = $lastResult < 0 ? 0 : $lastResult;
+        $response->headers->set('Content-Range', "$start-$lastResult/" . count($collection));
+
+        return $response;
+    }
+
+     private function getKeywordRepository()
+    {
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: check your indentation, php-cs-fixer can fix this kind of stuff too.

File: Rest/Controller/KeywordController.php [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\BadRequestHttpException;
+use Symfony\Component\Validator\Constraints as Assert;
+
+use BackBee\Exception\InvalidArgumentException;
+use BackBee\NestedNode\MediaFolder;
+use BackBee\Rest\Controller\Annotations as Rest;
+use BackBee\Rest\Patcher\EntityPatcher;
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: YAGNI, remove all unused use statments please.

Or use php-cs-fixer tool to remove it

File: Rest/Controller/KeywordController.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        $response->headers->set('Content-Range', "$start-$lastResult/" . count($collection));
+
+        return $response;
+    }
+
+     private function getKeywordRepository()
+    {
+        return $this->getEntityManager()->getRepository('BackBee\NestedNode\KeyWord');
+    }
+
```
Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: remove also all unused lines

**Comment by mickaelandrieu** [comment]
Also, this need tests as this is part of REST api.

File: Rest/Controller/KeywordController.php (hunk 8 above)
Comment (hbaptiste) [inline Rest/Controller/KeywordController.php]: what do you mean?

Comment (mickaelandrieu) [inline Rest/Controller/KeywordController.php]: What I mean is you have to create your array in KeywordRepository class and not in your controller, what do you think ?

**Comment by mickaelandrieu** [comment]
thank you @hbaptiste, good job !

**Pull request merged**